extension/src/functions.rs

```diff
@@ -133,6 +133,32 @@ impl<M: Matches, C, R, D, F, RC, E, L> Matches for ReadState<M, C, R, D, F, RC,
 	}
 }
 
+/// A function that filters the provided input. Only for testing purpose.
+/// Returns supplied contract value as 1 if allowed, 0 if not allowed.
+struct InputFilter<M, C, F>(PhantomData<(M, C, F)>);
```

Comment:
Thank to the types you create in mock.rs (see other comment), why create an entire new function?

Creates a lot of noise and a lot more separate tests. Perhaps we can refactor the existing tests and use some kind of `for` loop where we provide the filter and execute all the same tests. Like you did in `decoding` with all the enums. 

The tests on itself are not big so I rather have extensive tests in stead of a lot of small ones